DEFI/LiquidityPool/certora/specs/Full.spec

```diff
@@ -0,0 +1,448 @@
+    function flashLoan(address, uint256)                 external;
+
+    function underlying.balanceOf(address)               external returns(uint256) envfree;
+    function underlying.allowance(address, address)      external returns(uint256) envfree;
+    function underlying.totalSupply()                    external returns(uint256) envfree;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Reentrancy ghost and hook                                                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+ghost uint256 ghostReentrancyStatus;
+
+hook Sload uint256 status currentContract._status STORAGE {
+    require ghostReentrancyStatus == status;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Deposite                                                                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule depositIntegrity(env e){
+
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 amount;
+    uint256 clientBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    
+    deposit(e, amount);
+
+    uint256 clientBalanceAfter = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesAfter = balanceOf(e.msg.sender);
+
+    assert (amount == 0) => (clientBalanceBefore == clientBalanceAfter) && (clientSharesBefore == clientSharesAfter);
+    assert (amount > 0) => (clientBalanceBefore > clientBalanceAfter) && (clientSharesBefore < clientSharesAfter);
+}
+
+rule depositRevertConditions(env e){
+
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 amount;
+    uint256 clientBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    
+    bool underFlow = clientBalanceBefore - amount < 0;
+    bool emptyPool = totalSupply() == 0 || underlying.balanceOf(currentContract) == 0;
+    bool clientSharesOverflow = (clientSharesBefore + amount > max_uint256 && emptyPool) || clientSharesBefore + amountToShares(amount) > max_uint256;
+    bool totalSharesOverflow = totalSupply() + amountToShares(amount) > max_uint256;
+    bool contractUnderlyingOverflow = underlying.balanceOf(currentContract) + amount > max_uint256;
+    bool overflow =  clientSharesOverflow || totalSharesOverflow || contractUnderlyingOverflow;
+    bool payable = e.msg.value != 0;
+    bool reentrancy = ghostReentrancyStatus == 2;
```

Comment:
better to use definition. e.g. definition lock_on() == 2

Reply:
Done